### PR TITLE
feat: show span as soon as they arrive

### LINKED ISSUE #2350: [BUG] Show spans to the UI as soon as they arrive
**Describe the bug**
Right now phoenix doesn't show spans that are not a part of a complete trace (so that we can compute things like latency)

However as a user that is debugging a system - I want to see the spans as soon as they arrive and even if the parent span is not completed. This helps with debugging - notably LLM calls can fail due to context window overflows, variable mismatches etc.

I don't want to wait or depend on the parent span to complete before I see the output in the UI.

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #2350 

Spans with incomplete traces will only show up in the Spans tab, and not the Trace tab.

Screenshot 2024-02-22 at 3 42 23 PM

## First screenshot below shows the partial trace.
- The header is is a just placeholder because the true root span is not present.
- All partitions of the tree are shown at the same time. (There are two partitions in the screenshot.)
Screenshot 2024-02-22 at 3 30 10 PM

## This next screenshot shows how it looks when the trace is finally complete.
- Note that the header is now shown because it displays values based on the root span (i.e. the span whose parent is null)
Screenshot 2024-02-22 at 12 58 54 PM

## Here's the code used for testing.
- A breakpoint can be placed e.g. at the print() statement.
- Uncomment the raise statement to test exception handling.
```python
import openai
from openinference.instrumentation.openai import OpenAIInstrumentor
from opentelemetry import trace as trace_api
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk import trace as trace_sdk
from opentelemetry.sdk.trace.export import SimpleSpanProcessor

endpoint = "http://127.0.0.1:6006/v1/traces"
tracer_provider = trace_sdk.TracerProvider()
tracer_provider.add_span_processor(SimpleSpanProcessor(OTLPSpanExporter(endpoint, timeout=10000)))

OpenAIInstrumentor().instrument(tracer_provider=tracer_provider)


if __name__ == "__main__":
    client = openai.OpenAI()
    tracer = trace_api.get_tracer(__name__)
    with tracer.start_as_current_span("root") as span:
        response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=[{"role": "user", "content": "Write a poem."}],
            max_tokens=20,
        )
        with tracer.start_as_current_span("parent") as span:
            response = client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[{"role": "user", "content": "Write a haiku."}],
                max_tokens=20,
            )
            # raise ValueError("This is a test error")
            response = client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[{"role": "user", "content": "Write a sonnet."}],
                max_tokens=20,
            )
        print()
```

### CODE REVIEW

File: src/phoenix/core/traces.py [hunk 1]
```diff
@@ -108,22 +118,20 @@ def __init__(self) -> None:
...
-        self._traces: DefaultDict[TraceID, List[SpanID]] = defaultdict(list)
-        self._child_span_ids: DefaultDict[SpanID, Set[ChildSpanID]] = defaultdict(set)
+        self._traces: DefaultDict[TraceID, Set[ReadableSpan]] = defaultdict(set)
+        self._child_spans: DefaultDict[SpanID, Set[ReadableSpan]] = defaultdict(set)
         self._orphan_spans: DefaultDict[ParentSpanID, List[otlp.Span]] = defaultdict(list)
         self._num_documents: DefaultDict[SpanID, int] = defaultdict(int)
-        self._start_time_sorted_span_ids: SortedKeyList[SpanID] = SortedKeyList(
-            key=lambda span_id: self._spans[span_id].start_time,
+        self._start_time_sorted_spans: SortedKeyList[ReadableSpan] = SortedKeyList(
+            key=lambda span: span.start_time,
```
Comment (RogerHYang) [inline src/phoenix/core/traces.py:126]: this is a refactor that removes the extra step of looking up a span by its id

**Comment by RogerHYang** [comment]
Here's the code used for testing. A breakpoint can be placed e.g. at the print() statement.
```python
import openai
from openinference.instrumentation.openai import OpenAIInstrumentor
from opentelemetry import trace as trace_api
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk import trace as trace_sdk
from opentelemetry.sdk.trace.export import ConsoleSpanExporter, SimpleSpanProcessor

endpoint = "http://127.0.0.1:6006/v1/traces"
tracer_provider = trace_sdk.TracerProvider()
tracer_provider.add_span_processor(SimpleSpanProcessor(OTLPSpanExporter(endpoint, timeout=10000)))
tracer_provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))
trace_api.set_tracer_provider(tracer_provider)

OpenAIInstrumentor().instrument()


if __name__ == "__main__":
    client = openai.OpenAI()
    template = "Write a {item}."
    item = {"item": "haiku"}
    tracer = trace_api.get_tracer(__name__)
    with tracer.start_as_current_span("root") as span:
        response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=[{"role": "user", "content": "Write a poem."}],
            max_tokens=20,
        )
        with tracer.start_as_current_span("parent") as span:
            response = client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[{"role": "user", "content": "Write a haiku."}],
                max_tokens=20,
            )
            response = client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[{"role": "user", "content": "Write a sonnet."}],
                max_tokens=20,
            )
        print()
```

File: app/src/pages/trace/TracePage.tsx [hunk 2]
```diff
@@ -214,7 +214,10 @@ export function TracePage() {
         }
       }
     `,
-    { traceId: traceId as string }
+    { traceId: traceId as string },
+    {
+      fetchPolicy: "network-only",
+    }
```
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:220]: Do you still need this? Alternatively you can use fetch-and-network https://relay.dev/docs/guided-tour/reusing-cached-data/fetch-policies/

File: app/src/pages/trace/TracePage.tsx [hunk 3]
```diff
@@ -226,9 +229,6 @@ export function TracePage() {
     return spansList.find((span) => span.parentId == null);
   }, [spansList]);
 
-  if (rootSpan == null) {
-    throw new Error("rootSpan is required to view a trace");
-  }
```
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:231]: don't you want to ensure there is a root span?

File: app/src/pages/trace/TracePage.tsx [hunk 4]
```diff
@@ -244,7 +244,7 @@ export function TracePage() {
             flex-direction: column;
           `}
         >
-          <TraceHeader rootSpan={rootSpan} />
+          {rootSpan ? <TraceHeader rootSpan={rootSpan} /> : null}
```
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:247]: should we just use the first root span for now?

**Review by mikeldking — APPROVED** [review]

File: app/src/pages/trace/TracePage.tsx [hunk 5]
```diff
@@ -282,8 +282,12 @@ export function TracePage() {
   );
 }
 
-function TraceHeader({ rootSpan }: { rootSpan: Span }) {
-  const { latencyMs, statusCode, spanEvaluations } = rootSpan;
+function TraceHeader({ rootSpan }: { rootSpan: Span | null | undefined }) {
```
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:285]:
```suggestion
function TraceHeader({ rootSpan }: { rootSpan?: Span | null  }) {
```

**Review by mikeldking — APPROVED** [review]

**Pull request merged**